File: nekoshlink/shlink-security/src/main/java/org/nekosoft/shlink/sec/ShlinkSecurityConfiguration.java
package org.nekosoft.shlink.sec;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.autoconfigure.condition.ConditionalOnNotWebApplication;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.access.hierarchicalroles.RoleHierarchy;
import org.springframework.security.access.hierarchicalroles.RoleHierarchyImpl;
import org.springframework.security.authentication.AnonymousAuthenticationProvider;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.ProviderManager;
import org.springframework.security.authentication.dao.DaoAuthenticationProvider;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.oauth2.jwt.JwtDecoders;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationConverter;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationProvider;
import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter;

@Configuration
public class ShlinkSecurityConfiguration {

    public static final String VERSION_STRING = "1";

    @Bean
    public PasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }

    @Bean
    @ConditionalOnNotWebApplication
    public JwtAuthenticationProvider jwtAuthProvider(
            @Value("${spring.security.oauth2.resourceserver.jwt.issuer-uri}") String issuerUri,
            JwtAuthenticationConverter converter
            ) {
        JwtAuthenticationProvider provider = new JwtAuthenticationProvider(
                JwtDecoders.fromIssuerLocation(issuerUri)
        );
        provider.setJwtAuthenticationConverter(converter);
        return provider;
    }

    @Bean
    public JwtAuthenticationConverter jwtAuthenticationConverter() {
        JwtGrantedAuthoritiesConverter jwtGrantedAuthoritiesConverter = new JwtGrantedAuthoritiesConverter();
        jwtGrantedAuthoritiesConverter.setAuthoritiesClaimName("nkshlink-roles");
        jwtGrantedAuthoritiesConverter.setAuthorityPrefix("ROLE_");
        JwtAuthenticationConverter jwtAuthenticationConverter = new JwtAuthenticationConverter();
        jwtAuthenticationConverter.setJwtGrantedAuthoritiesConverter(jwtGrantedAuthoritiesConverter);
        return jwtAuthenticationConverter;
    }

    @Bean
    public DaoAuthenticationProvider userProvider(UserDetailsService userService, PasswordEncoder pwdEncoder) {
        DaoAuthenticationProvider userProvider = new DaoAuthenticationProvider();
        userProvider.setUserDetailsService(userService);
        userProvider.setPasswordEncoder(pwdEncoder);
        return userProvider;
    }

    @Bean
    @ConditionalOnNotWebApplication
    public AuthenticationManager authManager(
            ApiKeyAuthenticationProvider apiKeyProvider,
            JwtAuthenticationProvider oauth2Provider,
            DaoAuthenticationProvider userProvider
    ) {
        return new ProviderManager(
                apiKeyProvider,
                oauth2Provider,
                userProvider,
                new AnonymousAuthenticationProvider("NekoShlink")
        );
    }

    @Bean
    public RoleHierarchy roleHierarchy() {
        RoleHierarchyImpl hierarchy = new RoleHierarchyImpl();
        hierarchy.setHierarchy("""
            ROLE_Admin > ROLE_Editor
            ROLE_Editor > ROLE_Viewer
        """.stripIndent());
        return hierarchy;
    }


}

File: nekoshlink/shlink-security/src/main/kotlin/org/nekosoft/shlink/sec/ShlinkSecurityConfiguration.kt
package org.nekosoft.shlink.sec

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.authentication.AnonymousAuthenticationProvider
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.ProviderManager
import org.springframework.security.authentication.dao.DaoAuthenticationProvider
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.provisioning.InMemoryUserDetailsManager
import org.springframework.security.provisioning.UserDetailsManager

@Configuration
class ShlinkSecurityConfiguration {

    @Bean
    fun userDetailsService(): UserDetailsService {
        val adminAuthority = HashSet<GrantedAuthority>()
        adminAuthority.add(SimpleGrantedAuthority("ROLE_Admin"))
        adminAuthority.add(SimpleGrantedAuthority("ROLE_User"))
        adminAuthority.add(SimpleGrantedAuthority("ROLE_Anyone"))
        val userAuthority = HashSet<GrantedAuthority>()
        userAuthority.add(SimpleGrantedAuthority("ROLE_User"))
        userAuthority.add(SimpleGrantedAuthority("ROLE_Anyone"))
        val guestAuthority = HashSet<GrantedAuthority>()
        guestAuthority.add(SimpleGrantedAuthority("ROLE_Anyone"))
        return InMemoryUserDetailsManager(
            User("admin", "{noop}password1", adminAuthority),
            User("user", "{noop}password2", userAuthority),
            User("guest", "{noop}password3", guestAuthority),
        )
    }

    @Bean
    fun userProvider(detailsService: UserDetailsService): DaoAuthenticationProvider {
        val userProvider = DaoAuthenticationProvider()
        userProvider.setUserDetailsService(detailsService)
        return userProvider
    }

    @Bean
    fun authManager(
        apiKeyProvider: ApiKeyAuthenticationProvider,
        userProvider: DaoAuthenticationProvider,
    ): AuthenticationManager = ProviderManager(
        apiKeyProvider,
        userProvider,
        AnonymousAuthenticationProvider("NekoShlink")
    )

}